File: Facade/Facade.ts
class Engine {
  public start(): void {
    console.log("Motor ligado");
  }

  public stop(): void {
    console.log("Motor desligado");
  }
}

class Lights {
  public turnOn(): void {
    console.log("Faróis ligados");
  }

  public turnOff(): void {
    console.log("Faróis desligados");
  }
}

class FuelSystem {
  public pumpFuel(): void {
    console.log("Bomba de combustível ligada");
  }

  public stopPumpingFuel(): void {
    console.log("Bomba de combustível desligada");
  }
}

class CarFacade {
  private readonly _engine: Engine;
  private readonly _lights: Lights;
  private readonly _fuelSystem: FuelSystem;

  constructor() {
    this._engine = new Engine();
    this._lights = new Lights();
    this._fuelSystem = new FuelSystem();
  }

  public startCar(): void {
    this._engine.start();
    this._lights.turnOn();
    this._fuelSystem.pumpFuel();
    console.log("Carro pronto para uso");
  }

  public stopCar(): void {
    this._fuelSystem.stopPumpingFuel();
    this._lights.turnOff();
    this._engine.stop();
    console.log("Carro desligado");
  }
}

File: Facade/Facade.cs
class Engine
{
  public void Start()
  {
    Console.WriteLine("Motor ligado");
  }

  public void Stop()
  {
    Console.WriteLine("Motor desligado");
  }
}

class Lights
{
  public void TurnOn()
  {
    Console.WriteLine("Faróis ligados");
  }

  public void TurnOff()
  {
    Console.WriteLine("Faróis desligados");
  }
}

class FuelSystem
{
  public void PumpFuel()
  {
    Console.WriteLine("Bomba de combustível ligada");
  }

  public void StopPumpingFuel()
  {
    Console.WriteLine("Bomba de combustível desligada");
  }
}

class CarFacade
{
  private readonly Engine _engine;
  private readonly Lights _lights;
  private readonly FuelSystem _fuelSystem;

  public CarFacade()
  {
    _engine = new Engine();
    _lights = new Lights();
    _fuelSystem = new FuelSystem();
  }

  public void StartCar()
  {
    _engine.Start();
    _lights.TurnOn();
    _fuelSystem.PumpFuel();
    Console.WriteLine("Carro pronto para uso");
  }

  public void StopCar()
  {
    _fuelSystem.StopPumpingFuel();
    _lights.TurnOff();
    _engine.Stop();
    Console.WriteLine("Carro desligado");
  }
}
